src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.netcore.cs

```diff
@@ -1220,102 +1222,892 @@ private static float MinMaxCore<TMinMaxOperator>(ReadOnlySpan<float> x)
+                                vector4 = TUnaryOperator.Invoke(Vector128.Load(xPtr + (uint)(Vector128<float>.Count * 3)));
+
+                                vector1.Store(dPtr + (uint)(Vector128<float>.Count * 0));
+                                vector2.Store(dPtr + (uint)(Vector128<float>.Count * 1));
+                                vector3.Store(dPtr + (uint)(Vector128<float>.Count * 2));
+                                vector4.Store(dPtr + (uint)(Vector128<float>.Count * 3));
+
+                                // We load, process, and store the next four vectors
+
+                                vector1 = TUnaryOperator.Invoke(Vector128.Load(xPtr + (uint)(Vector128<float>.Count * 4)));
+                                vector2 = TUnaryOperator.Invoke(Vector128.Load(xPtr + (uint)(Vector128<float>.Count * 5)));
+                                vector3 = TUnaryOperator.Invoke(Vector128.Load(xPtr + (uint)(Vector128<float>.Count * 6)));
+                                vector4 = TUnaryOperator.Invoke(Vector128.Load(xPtr + (uint)(Vector128<float>.Count * 7)));
+
+                                vector1.Store(dPtr + (uint)(Vector128<float>.Count * 4));
+                                vector2.Store(dPtr + (uint)(Vector128<float>.Count * 5));
+                                vector3.Store(dPtr + (uint)(Vector128<float>.Count * 6));
+                                vector4.Store(dPtr + (uint)(Vector128<float>.Count * 7));
+
+                                // We adjust the source and destination references, then update
+                                // the count of remaining elements to process.
+
+                                xPtr += (uint)(Vector128<float>.Count * 8);
+                                dPtr += (uint)(Vector128<float>.Count * 8);
+
+                                remainder -= (uint)(Vector128<float>.Count * 8);
+                            }
+                        }
+
+                        // Adjusting the refs here allows us to avoid pinning for very small inputs
+
+                        xRef = ref *xPtr;
+                        dRef = ref *dPtr;
                     }
-                    while (i <= oneVectorFromEnd);
+                }
 
-                    // Handle any remaining elements with a final vector.
-                    if (i != x.Length)
+                // Process the remaining [Count, Count * 8] elements via a jump table
+                //
+                // Unless the original length was an exact multiple of Count, then we'll
+                // end up reprocessing a couple elements in case 1 for end. We'll also
+                // potentially reprocess a few elements in case 0 for beg, to handle any
+                // data before the first aligned address.
+
+                nuint endIndex = remainder;
+                remainder = (remainder + (uint)(Vector128<float>.Count - 1)) & (nuint)(-Vector128<float>.Count);
+
+                switch (remainder / (uint)(Vector128<float>.Count))
```

Comment:
Is https://github.com/dotnet/runtime/pull/93346/files#r1355278777 relevant here at all?

Reply:
I don't think so. Native memcpy algorithms have similar jump tables, explicitly, because the number of remaining is unpredictable.

That is, the entire point is the number of remaining blocks can't be predicted. But we need to have handling for each. Looping is undesirable because the branch predictor gets it wrong (both by default because loops are predicted to be taken at least 6-8 times, but also over time because the number of iterations is not consistent).

A fallthrough jump table, such as this, allows us to have consistent jump target addresses, where we use the index as an offset to jump. Even though the branch itself is unpredictable it is unconditional and the destinations are often already decoded or partially decoded due to being immediately after the jump which allows for reduce cost. This makes it better than many conditional branches which can likewise not be predicted.

Using such jump tables is actually explicitly covered in the Intel/AMD optimization manuals, typically as part of the vectorization logic. The current May 2023 copy of `Intel® 64 and IA-32 Architectures Optimization Reference Manual` covers this:
> After an initial unaligned
store, and adjusting the destination pointer to be 32-byte aligned, the residual phase follows the same
consideration as described in Section 15.16.3.1, which may employ a large jump table to handle each
residual value scenario with minimal branching, depending on the amount of unrolled 32B-aligned
stores. The main loop is a simple YMM register to 32-byte-aligned store operation, which can deliver
close to 30 bytes per cycle for lengths more than a thousand byte. The limiting factor here is due to each
256-bit VMOVDQA store consists of a store_address and a store_data micro-op flow. Only port 4 is available
to dispatch the store_data micro-op each cycle.

Among other such references
> On Ivy Bridge microarchitecture, a REP MOVSB implementation of memcpy can achieve throughput at
slightly better than the 128-bit SIMD implementation when copying thousands of bytes. However, if the
size of copy operation is less than a few hundred bytes, the REP MOVSB approach is less efficient than the
explicit residual copy technique described in phase 2 of Section 15.16.3.1. This is because handling 1-
127 residual copy length (via jump table or switch/case, and is done before the main loop) plus one or
two 8x16B iterations incurs less branching overhead than the hardware provided micro-op flows. For the
grueling implementation details of 128-bit SIMD implementation of memcpy(), one can look up from the
archived sources of open source library such as GLibC.

or the discussions on how the instruction cache works
> For code entering the Decoded ICache, each unconditional branch is the last micro-op occupying a
Decoded ICache Way. Therefore, only three unconditional branches per a 32 byte aligned chunk can
enter the Decoded ICache.
Unconditional branches are frequent in jump tables and switch declarations. Below are examples for
these constructs, and methods for writing them so that they fit in the Decoded ICache.
Compilers create jump tables for C++ virtual class methods or DLL dispatch tables. Each unconditional
branch consumes five bytes; therefore up to seven of them can be associated with a 32-byte chunk. Thus
jump tables may not fit in the Decoded ICache if the unconditional branches are too dense in each
32Byte-aligned chunk. This can cause performance degradation for code executing before and after the
branch table.
The solution is to add multi-byte NOP instructions among the branches in the branch table. This may
increases code size and should be used cautiously. However, these NOPs are not executed and therefore
have no penalty in later pipe stages.
Switch-Case constructs represents a similar situation. Each evaluation of a case condition results in an
unconditional branch. The same solution of using multi-byte NOP can apply for every three consecutive
unconditional branches that fits inside an aligned 32-byte chunk.